src/crypto/crypto_util.cc

```diff
@@ -125,7 +125,6 @@ bool ProcessFipsOptions() {
     OSSL_PROVIDER* fips_provider = OSSL_PROVIDER_load(nullptr, "fips");
     if (fips_provider == nullptr)
       return false;
-    OSSL_PROVIDER_unload(fips_provider);
```

Comment:
@danbev may remember why the load/unload. From the PR my read is that it was to be able to through an error during options processing, but not have the provider loaded unless it was used.

Looking at the other branch I'm wondering if this would be consistent with what we have in the non >= section

```C++
#if OPENSSL_VERSION_MAJOR >= 3
  if (!EVP_default_properties_is_fips_enabled(nullptr)) {
    OSSL_PROVIDER* fips_provider = OSSL_PROVIDER_load(nullptr, "fips");
    if (fips_provider == nullptr)
      return false;
    OSSL_PROVIDER_unload(fips_provider);

    return EVP_default_properties_enable_fips(nullptr, 1) &&
    EVP_default_properties_is_fips_enabled(nullptr);
  }
```

I think you mentioned to me that doing that still fails.  If so, then I wonder why if FIPs is already enabled on the base system, then why the appropriate provider is not already loaded. It would be good to confirm if its expected that you still need to load the provider even if the system has enabled FIPs to make sure the correct provider is being used.  

@richardlau I know you have also look a bit at how FIPs and providers work so I'd be interested in any insights you have as well.

Reply:
Yes, according to my testing just having the guard there does not help *on non-FIPS system*! I did not test it yet on FIPS system, as in my opinion coredump should not have happen in any system configuration.

---

The issue might be in getting confused on what the `EVP_…` functions actually do – according to manual page, they do not check the provider status at all:
```
EVP_default_properties_enable_fips() sets the 'fips=yes' to be a default property if enable is non zero, otherwise it clears 'fips' from the default property query for the given libctx. It merges the fips default property query with any existing query strings that have been set via EVP_set_default_properties().

EVP_default_properties_is_fips_enabled() indicates if 'fips=yes' is a default property for the given libctx.
```

I'm very much guessing in the dark here, but I suspect setting the EVP property `fips=yes` does not load the provider, nor does it indicate it was ever loaded – it just indicate you want to use it by default.

Reply:
> 
> @richardlau I know you have also look a bit at how FIPs and providers work so I'd be interested in any insights you have as well.

FIPS support in OpenSSL 3 is completely different from what came before and works with the (new for OpenSSL 3) Provider model. This model is a lot more generic -- you may have any number of providers loaded (aka available) and can change the loaded providers at runtime (through `OSSL_PROVIDER_load`/`OSSL_PROVIDER_unload`). Since it's generic the providers themselves can be named anything, i.e. a FIPS provider does not need to be named "fips" (which means the code in Node.js being queried is overly specific and probably does need to change).

So what does "fips mode" mean in the provider model? When fetching an algorithm in OpenSSL the algorithm may be implemented by more than one loaded provider. If that is the case it is undefined which implemented is used unless the fetch was constrained by an property query, e.g. `fips=yes` -- note this may still match more than one provider if those providers set `fips=yes` on their algorithm implementations. You can limit to a specifically named provided by property (e.g. `provider=fips`) but remember that OpenSSL puts no special meaning on the provider names themselves. The property query can either be set on every individual fetch call or as a default property (which applies to all fetch calls that don't specify their own). For OpenSSL 3 Node.js treats "fips mode" as the equivalent of having set a default query of `fips=yes`.

The genericness of the provider model coupled with the ability to dynamically change the loaded providers also means that it is possible that an algorithm fetch returns no matching implementations. Such a corner case was reported in https://github.com/nodejs/node/issues/46200 -- if `fips=yes` is set as the default property query and no implementations that have that property are available (e.g. no provider has been loaded that provides a compatible implementation) then in that reported case it was not possible to initialize the cryptographically secure pseudorandom number generator (CSPRNG).

With regards to the code being commented on, I think the intention seems to be to determine if it might be possible to load a provider called "fips" (i.e. the provider is on the filesystem and could be loaded). This is a highly specific assumption -- OpenSSL 3 makes no guarantee that a provider called "fips" actually implements any FIPS compatible algorithms -- so is probably a flawed assumption to be making. I believe OpenSSL's `OSSL_PROVIDER_load`/`OSSL_PROVIDER_unload` operates via reference counting (so the provider may not actually be unloaded if it had already been loaded prior to the `OSSL_PROVIDER_load` call on L125).